File: plugins/plugin-scratchpad/python/elizaos_plugin_scratchpad/actions/read.py
"""SCRATCHPAD_READ action - Read a specific scratchpad entry by ID."""

from __future__ import annotations

import logging
import re
from dataclasses import dataclass
from typing import TYPE_CHECKING

from elizaos_plugin_scratchpad.service import create_scratchpad_service
from elizaos_plugin_scratchpad.types import ScratchpadReadOptions

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State

logger = logging.getLogger(__name__)

EXTRACT_TEMPLATE = """Extract the scratchpad entry ID and optional line range from the user's message.

User message: {{text}}

Available scratchpad entries:
{{entries}}

Respond with XML containing:
- id: The ID of the scratchpad entry to read (required)
- from: Starting line number (optional)
- lines: Number of lines to read (optional)

<response>
<id>entry-id</id>
<from>1</from>
<lines>10</lines>
</response>"""


def _parse_xml_field(text: str, field_name: str) -> str | None:
    """Extract a field value from XML-like response text."""
    match = re.search(rf"<{field_name}>([\s\S]*?)</{field_name}>", text)
    return match.group(1).strip() if match else None


@dataclass
class ReadResult:
    """Result of the SCRATCHPAD_READ action."""

    success: bool
    text: str
    error: str | None = None


async def _extract_read_info(
    runtime: IAgentRuntime,
    message: Memory,
    available_entries: str,
) -> dict[str, str | int] | None:
    """Use the LLM to extract the entry ID and optional line range."""
    message_text = message.content.text if message.content else ""
    prompt = EXTRACT_TEMPLATE.replace("{{text}}", message_text).replace(
        "{{entries}}", available_entries
    )

    result = await runtime.use_model("TEXT_SMALL", {"prompt": prompt, "stopSequences": []})
    result_str = str(result)
    logger.debug("[ScratchpadRead] Extract result: %s", result_str)

    entry_id = _parse_xml_field(result_str, "id")
    if not entry_id:
        logger.error("[ScratchpadRead] Failed to extract valid read info")
        return None

    info: dict[str, str | int] = {"id": entry_id}
    from_str = _parse_xml_field(result_str, "from")
    if from_str:
        try:
            info["from_line"] = int(from_str)
        except ValueError:
            pass
    lines_str = _parse_xml_field(result_str, "lines")
    if lines_str:
        try:
            info["lines"] = int(lines_str)
        except ValueError:
            pass

    return info


async def handle_scratchpad_read(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
    options: HandlerOptions | None = None,
    callback: HandlerCallback | None = None,
    responses: list[Memory] | None = None,
) -> ReadResult | None:
    """Handle the SCRATCHPAD_READ action."""
    service = create_scratchpad_service(runtime)

    entries = await service.list()
    entries_context = "\n".join(f'- {e.id}: "{e.title}"' for e in entries)

    if not entries:
        msg = "There are no scratchpad entries to read. You can create one first."
        if callback:
            await callback(
                {
                    "text": msg,
                    "actions": ["SCRATCHPAD_READ_EMPTY"],
                    "source": message.content.source if message.content else None,
                }
            )
        return ReadResult(success=False, text="No entries available")

    read_info = await _extract_read_info(runtime, message, entries_context)

    if not read_info:
        msg = f"I couldn't determine which note to read. Available entries:\n{entries_context}"
        if callback:
            await callback(
                {
                    "text": msg,
                    "actions": ["SCRATCHPAD_READ_FAILED"],
                    "source": message.content.source if message.content else None,
                }
            )
        return ReadResult(success=False, text="Failed to extract read info")

    try:
        entry_id = str(read_info["id"])
        from_line = read_info.get("from_line")
        num_lines = read_info.get("lines")
        read_opts = ScratchpadReadOptions(
            from_line=int(from_line) if from_line is not None else None,
            lines=int(num_lines) if num_lines is not None else None,
        )

        entry = await service.read(entry_id, read_opts)

        line_info = ""
        if from_line is not None:
            end_line = (int(from_line) or 1) + (int(num_lines) if num_lines is not None else 10)
            line_info = f" (lines {from_line}-{end_line})"

        success_msg = f"**{entry.title}**{line_info}\n\n{entry.content}"

        if callback:
            await callback(
                {
                    "text": success_msg,
                    "actions": ["SCRATCHPAD_READ_SUCCESS"],
                    "source": message.content.source if message.content else None,
                }
            )

        return ReadResult(success=True, text=success_msg)

    except Exception as exc:
        error_msg = f"Failed to read the note: {exc}"
        logger.error("[ScratchpadRead] Error: %s", exc)
        if callback:
            await callback(
                {
                    "text": error_msg,
                    "actions": ["SCRATCHPAD_READ_FAILED"],
                    "source": message.content.source if message.content else None,
                }
            )
        return ReadResult(success=False, text="Failed to read scratchpad entry", error=str(exc))


async def validate_scratchpad_read(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> bool:
    """Always validate - selection is handled in the handler."""
    return True


SCRATCHPAD_READ_ACTION: dict[str, object] = {
    "name": "SCRATCHPAD_READ",
    "similes": [
        "READ_NOTE",
        "GET_NOTE",
        "VIEW_NOTE",
        "SHOW_NOTE",
        "OPEN_SCRATCHPAD",
    ],
    "description": "Read a specific scratchpad entry by its ID, with optional line range.",
    "examples": [
        [
            {
                "name": "{{name1}}",
                "content": {"text": "Read my note about project deadlines"},
            },
            {
                "name": "{{name2}}",
                "content": {
                    "text": "**Project Deadline**\n\nThe deadline is March 15th.",
                    "actions": ["SCRATCHPAD_READ_SUCCESS"],
                },
            },
        ],
    ],
}

File: packages/core/src/features/advanced-capabilities/clipboard/actions/read.ts
import {
	type Action,
	type HandlerCallback,
	type HandlerOptions,
	type IAgentRuntime,
	logger,
	type Memory,
	type State,
} from "../../../../types/index.ts";
import { hasActionContextOrKeyword } from "../../../../utils/action-validation.ts";
import { createClipboardService } from "../services/clipboardService.ts";
import { requireActionSpec } from "../specs.ts";

interface ReadInput {
	id: string;
	from?: number;
	lines?: number;
}

function isValidReadInput(obj: Record<string, unknown>): boolean {
	return typeof obj.id === "string" && obj.id.length > 0;
}

function readParams(options?: HandlerOptions): Record<string, unknown> {
	return options?.parameters && typeof options.parameters === "object"
		? (options.parameters as Record<string, unknown>)
		: {};
}

function optionalNumber(value: unknown): number | undefined {
	if (typeof value === "number" && Number.isFinite(value)) return value;
	if (typeof value === "string" && value.trim()) {
		const parsed = Number(value);
		return Number.isFinite(parsed) ? parsed : undefined;
	}
	return undefined;
}

function extractReadInfo(
	message: Memory,
	options?: HandlerOptions,
): ReadInput | null {
	const params = readParams(options);
	const raw = {
		id: params.id ?? message.content.id ?? message.content.entryId,
		from: params.from ?? message.content.from,
		lines: params.lines ?? message.content.lines,
	};

	if (!isValidReadInput(raw)) {
		logger.error("[ClipboardRead] Failed to extract valid read info");
		return null;
	}

	return {
		id: String(raw.id),
		from: optionalNumber(raw.from),
		lines: optionalNumber(raw.lines),
	};
}

const spec = requireActionSpec("CLIPBOARD_READ");

export const clipboardReadAction: Action = {
	name: spec.name,
	contexts: ["files", "knowledge", "agent_internal"],
	roleGate: { minRole: "ADMIN" },
	similes: spec.similes ? [...spec.similes] : [],
	description: spec.description,

	validate: async (
		_runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		options?: HandlerOptions,
	): Promise<boolean> => {
		const __avParams = readParams(options);
		if (isValidReadInput(__avParams)) {
			return true;
		}
		return hasActionContextOrKeyword(message, state, {
			contexts: ["files", "knowledge", "agent_internal"],
			keywords: ["clipboard", "read note", "open note", "show note"],
		});
	},

	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		_stateFromTrigger: State | undefined,
		_options: HandlerOptions | undefined,
		callback?: HandlerCallback,
		_responses?: Memory[],
	) => {
		const service = createClipboardService(runtime);

		// Get list of available entries for context
		const entries = await service.list();
		const entriesContext = entries
			.map((e) => `- ${e.id}: "${e.title}"`)
			.join("\n");

		if (entries.length === 0) {
			if (callback) {
				await callback({
					text: "There are no clipboard entries to read. You can create one first.",
					actions: ["CLIPBOARD_READ_EMPTY"],
					source: message.content.source,
				});
			}
			return { success: false, text: "No entries available" };
		}

		const readInfo = extractReadInfo(message, _options);

		if (!readInfo) {
			if (callback) {
				await callback({
					text: `I couldn't determine which note to read. Available entries:\n${entriesContext}`,
					actions: ["CLIPBOARD_READ_FAILED"],
					source: message.content.source,
				});
			}
			return { success: false, text: "Failed to extract read info" };
		}

		try {
			const entry = await service.read(readInfo.id, {
				from: readInfo.from,
				lines: readInfo.lines,
			});

			const lineInfo =
				readInfo.from !== undefined
					? ` (lines ${readInfo.from}-${(readInfo.from ?? 1) + (readInfo.lines ?? 10)})`
					: "";

			const successMessage = `**${entry.title}**${lineInfo}\n\n${entry.content}`;

			if (callback) {
				await callback({
					text: successMessage,
					actions: ["CLIPBOARD_READ_SUCCESS"],
					source: message.content.source,
				});
			}

			return { success: true, text: successMessage, entry };
		} catch (error) {
			const errorMsg = error instanceof Error ? error.message : String(error);
			logger.error("[ClipboardRead] Error:", errorMsg);
			if (callback) {
				await callback({
					text: `Failed to read the note: ${errorMsg}`,
					actions: ["CLIPBOARD_READ_FAILED"],
					source: message.content.source,
				});
			}
			return { success: false, text: "Failed to read clipboard entry" };
		}
	},

	parameters: [
		{
			name: "id",
			description: "Clipboard entry ID to read.",
			required: true,
			schema: { type: "string" as const, minLength: 1 },
		},
		{
			name: "from",
			description: "Optional 1-based starting line number.",
			required: false,
			schema: { type: "number" as const, minimum: 1 },
		},
		{
			name: "lines",
			description: "Optional maximum number of lines to read.",
			required: false,
			schema: { type: "number" as const, minimum: 1 },
		},
	],
	examples: [],
};

export default clipboardReadAction;
